File: fpinjava-parent/fpinjava-lists-exercises/src/main/java/com/fpinjava/lists/exercise05_08/Product.java
package com.fpinjava.lists.exercise05_08;

import com.fpinjava.lists.exercise05_06.List;

public class Product {

	public static Double product(List<Double> doubles) {
		return doubles.isEmpty()
				? 1.0
				: doubles.head() == 0
				? 0.0
				: doubles.head() * product(doubles.tail());
	}
}

File: fpinjava-parent/fpinjava-lists-exercises/src/main/kotlin/com/fpinkotlin/lists/exercise05_08/Product.kt
package com.fpinkotlin.lists.exercise05_08

import com.fpinkotlin.lists.exercise05_06.List

fun product(doubles: List<Double>): Double =
		if (doubles.empty) 1.0
		else if (doubles.head == 0.0) 0.0
		else doubles.head * product(doubles.tail)
